File: app/src/main/java/com/example/uos_lms/core/domain/model/GradeScale.java
package com.example.uos_lms.core.domain.model;

import java.util.Arrays;
import java.util.List;

/**
 * HEC (Pakistan) style absolute grading scale - the only grading-scale
 * definition in the app. Bands are checked top-down against percentage, so
 * order matters (highest minPercent first).
 */
public final class GradeScale {

    private static final class Band {
        final double minPercent;
        final String letter;
        final double gpaPoint;

        Band(double minPercent, String letter, double gpaPoint) {
            this.minPercent = minPercent;
            this.letter = letter;
            this.gpaPoint = gpaPoint;
        }
    }

    private static final List<Band> BANDS = Arrays.asList(
            new Band(85.0, "A", 4.00),
            new Band(80.0, "A-", 3.66),
            new Band(75.0, "B+", 3.33),
            new Band(71.0, "B", 3.00),
            new Band(68.0, "B-", 2.66),
            new Band(64.0, "C+", 2.33),
            new Band(61.0, "C", 2.00),
            new Band(58.0, "C-", 1.66),
            new Band(54.0, "D+", 1.33),
            new Band(50.0, "D", 1.00),
            new Band(0.0, "F", 0.00)
    );

    private GradeScale() {
    }

    private static Band bandFor(double percentage) {
        for (Band band : BANDS) {
            if (percentage >= band.minPercent) return band;
        }
        return BANDS.get(BANDS.size() - 1);
    }

    public static String letterFor(double percentage) {
        return bandFor(percentage).letter;
    }

    public static double gpaPointFor(double percentage) {
        return bandFor(percentage).gpaPoint;
    }
}

File: src/services/gradeScale.js
// HEC (Pakistan) style absolute grading scale - the single source of truth for marks -> grade/
// gpa everywhere in the backend (per the spec: "not hardcoded in multiple places"). Ported
// exactly from the pre-migration app's GradeScale.java so results stay consistent with every
// grade already recorded there. Bands are checked top-down against percentage.
const BANDS = [
  { minPercent: 85.0, grade: 'A', gpa: 4.0 },
  { minPercent: 80.0, grade: 'A-', gpa: 3.66 },
  { minPercent: 75.0, grade: 'B+', gpa: 3.33 },
  { minPercent: 71.0, grade: 'B', gpa: 3.0 },
  { minPercent: 68.0, grade: 'B-', gpa: 2.66 },
  { minPercent: 64.0, grade: 'C+', gpa: 2.33 },
  { minPercent: 61.0, grade: 'C', gpa: 2.0 },
  { minPercent: 58.0, grade: 'C-', gpa: 1.66 },
  { minPercent: 54.0, grade: 'D+', gpa: 1.33 },
  { minPercent: 50.0, grade: 'D', gpa: 1.0 },
  { minPercent: 0.0, grade: 'F', gpa: 0.0 },
];

const FAILING_GRADE = 'F';

function bandFor(marks) {
  return BANDS.find((band) => marks >= band.minPercent) || BANDS[BANDS.length - 1];
}

/** marks: 0-100. Returns { grade, gpa, status }, never trust a client-submitted value. */
function gradeFor(marks) {
  const band = bandFor(marks);
  return { grade: band.grade, gpa: band.gpa, status: band.grade === FAILING_GRADE ? 'FAIL' : 'PASS' };
}

module.exports = { gradeFor, FAILING_GRADE };
